nm: Better handling for timeout

cathay4t opened the pull request:
When creating 1000 VLAN along with 1000 bridge using each VLAN,
NetworkManager might trigger two timeout:

 * The callback raises `Gio.IOErrorEnum.TIMED_OUT` error.
 * NetworkManager never call callback, nmstate idle check trigger the
   timeout.

To solve above issue:

 * Increase the nmstate idle check timeout to 5 minutes.

 * For actions like add profile, activate profile, add a
   fallback checker which check whether requested action is already
   finished using `GLib.timeout_source_new()`

 * When `Gio.IOErrorEnum.TIMED_OUT` happens, ignore the failure and wait
   fallback checker.

 * The fallback checker is only started 15 seconds after action started,
   so this does not impact small desire state.

Test results on RHEL 8.3 i7-8665U 2G RAM:

    13m3.895s to create 1000 VLAN and 1000 bridge over each VLAN.

ffmancera reviewed (approved):
Looks good to me, would it be possible to add the 1000 VLAN and 1000 Bridges over each VLAN scenario as a test? This way we are sure we are not reaching this failure anymore. I understand this test will take too much time..

cathay4t committed:
nm: Better handling for timeout

When creating 1000 VLAN along with 1000 bridge using each VLAN,
NetworkManager might trigger two timeout:

 * The callback raises `Gio.IOErrorEnum.TIMED_OUT` error.
 * NetworkManager never call callback, nmstate idle check trigger the
   timeout.

To solve above issue:

 * Increase the nmstate idle check timeout to 5 minutes.

 * For actions like add profile and activate profile, we add a
   fallback checker which check whether requested action is already
   finished using `GLib.timeout_source_new()`

 * When `Gio.IOErrorEnum.TIMED_OUT` happens, ignore the failure and wait
   fallback checker.

 * The fallback checker is only started 15 seconds after action started,
   so this does not impact small desire state.

Test results on RHEL 8.3 i7-8665U 2G RAM:

    13m3.895s to create 1000 VLAN and 1000 bridge over each VLAN.

Changed the integration test case to test 500 VLANs + 500 bridges.

Signed-off-by: Gris Ge <fge@redhat.com>
